commit 751b5736bb597210c5d9de40cb586d4b341bf29a
Author: Mark Wielaard <mark@klomp.org>
Date:   Tue Apr 9 16:09:50 2019 +0200

    More memcheck/tests/libstdc++.supp suppressions

diff --git a/memcheck/tests/libstdc++.supp b/memcheck/tests/libstdc++.supp
index fad537f..28ab182 100644
--- a/memcheck/tests/libstdc++.supp
+++ b/memcheck/tests/libstdc++.supp
@@ -49,6 +49,9 @@
 #    still reachable: 72,704 bytes in 1 blocks
 #         suppressed: 0 bytes in 0 blocks
 
+# All the following suppressions are variants of
+# _dl_init -> call_init which calls the DT_INIT_ARRAY functions
+
 {
    malloc-leaks-cxx-stl-string-classes
    Memcheck:Leak
@@ -61,6 +64,16 @@
    obj:*lib*/ld-2.*.so
 }
 {
+   malloc-leaks-cxx-stl-string-classes-2
+   Memcheck:Leak
+   match-leak-kinds: reachable
+   fun:malloc
+   obj:*lib*/libstdc++.so*
+   fun:call_init.part.0
+   fun:_dl_init
+   obj:*lib*/ld-2.*.so
+}
+{
   malloc-leaks-cxx-stl-string-classes-debug
   Memcheck:Leak
   match-leak-kinds: reachable
@@ -73,3 +86,13 @@
   fun:_dl_init
   obj:*lib*/ld-2.*.so
 }
+{
+   malloc-leaks-cxx-stl-string-classes-debug2
+   Memcheck:Leak
+   match-leak-kinds: reachable
+   fun:malloc
+   obj:*
+   fun:call_init
+   fun:_dl_init
+   fun:_dl_start_user
+}
