commit e63d90d66e8de6d6ec2c92e3b0a32cc165b83369
Author: Mark Wielaard <mark@klomp.org>
Date:   Wed Apr 10 01:22:37 2019 +0200

    dhat is broken on x86 (32bit)

diff --git a/dhat/dh_main.c b/dhat/dh_main.c
index 47a9327..a933328 100644
--- a/dhat/dh_main.c
+++ b/dhat/dh_main.c
@@ -794,6 +794,7 @@ void dh_handle_noninsn_write ( CorePart part, ThreadId tid,
 {
    switch (part) {
       case Vg_CoreSysCall:
+      case Vg_CoreClientReq:
          dh_handle_write(base, size);
          break;
       case Vg_CoreSignal:
