Only in valgrind-3.10.0.BETA2: autom4te.cache
diff -ur valgrind-3.10.0.BETA2.orig/config.h.in valgrind-3.10.0.BETA2/config.h.in
--- valgrind-3.10.0.BETA2.orig/config.h.in	2014-09-09 09:17:54.112654934 +0200
+++ valgrind-3.10.0.BETA2/config.h.in	2014-09-09 09:35:38.000000000 +0200
@@ -66,6 +66,9 @@
 /* Define to 1 if you're using glibc 2.2.x */
 #undef GLIBC_2_2
 
+/* Define to 1 if you're using glibc 2.20.x */
+#undef GLIBC_2_20
+
 /* Define to 1 if you're using glibc 2.3.x */
 #undef GLIBC_2_3
 
Only in valgrind-3.10.0.BETA2: config.h.in~
diff -ur valgrind-3.10.0.BETA2.orig/configure valgrind-3.10.0.BETA2/configure
--- valgrind-3.10.0.BETA2.orig/configure	2014-09-09 09:17:54.106654950 +0200
+++ valgrind-3.10.0.BETA2/configure	2014-09-09 09:35:43.280624261 +0200
@@ -6828,6 +6828,16 @@
 	DEFAULT_SUPP="glibc-2.34567-NPTL-helgrind.supp ${DEFAULT_SUPP}"
 	DEFAULT_SUPP="glibc-2.X-drd.supp ${DEFAULT_SUPP}"
 	;;
+     2.20)
+	{ $as_echo "$as_me:${as_lineno-$LINENO}: result: 2.20 family" >&5
+$as_echo "2.20 family" >&6; }
+
+$as_echo "#define GLIBC_2_20 1" >>confdefs.h
+
+	DEFAULT_SUPP="glibc-2.X.supp ${DEFAULT_SUPP}"
+	DEFAULT_SUPP="glibc-2.34567-NPTL-helgrind.supp ${DEFAULT_SUPP}"
+	DEFAULT_SUPP="glibc-2.X-drd.supp ${DEFAULT_SUPP}"
+	;;
      darwin)
 	{ $as_echo "$as_me:${as_lineno-$LINENO}: result: Darwin" >&5
 $as_echo "Darwin" >&6; }
diff -ur valgrind-3.10.0.BETA2.orig/configure.ac valgrind-3.10.0.BETA2/configure.ac
--- valgrind-3.10.0.BETA2.orig/configure.ac	2014-09-09 09:17:54.132654881 +0200
+++ valgrind-3.10.0.BETA2/configure.ac	2014-09-09 09:33:31.944021414 +0200
@@ -1008,6 +1008,13 @@
 	DEFAULT_SUPP="glibc-2.34567-NPTL-helgrind.supp ${DEFAULT_SUPP}"
 	DEFAULT_SUPP="glibc-2.X-drd.supp ${DEFAULT_SUPP}"
 	;;
+     2.20)
+	AC_MSG_RESULT(2.20 family)
+	AC_DEFINE([GLIBC_2_20], 1, [Define to 1 if you're using glibc 2.20.x])
+	DEFAULT_SUPP="glibc-2.X.supp ${DEFAULT_SUPP}"
+	DEFAULT_SUPP="glibc-2.34567-NPTL-helgrind.supp ${DEFAULT_SUPP}"
+	DEFAULT_SUPP="glibc-2.X-drd.supp ${DEFAULT_SUPP}"
+	;;
      darwin)
 	AC_MSG_RESULT(Darwin)
 	AC_DEFINE([DARWIN_LIBC], 1, [Define to 1 if you're using Darwin])
