diff --git a/VEX/priv/main_main.c b/VEX/priv/main_main.c
index 1253cf588..482047c7a 100644
--- a/VEX/priv/main_main.c
+++ b/VEX/priv/main_main.c
@@ -2163,11 +2163,6 @@ static void check_hwcaps ( VexArch arch, UInt hwcaps )
          if (have_fp16 != have_vfp16)
             invalid_hwcaps(arch, hwcaps,
                     "Mismatch detected between scalar and vector FP16 features.\n");
-         Bool have_rdm = ((hwcaps & VEX_HWCAPS_ARM64_RDM) != 0);
-         Bool have_atomics = ((hwcaps & VEX_HWCAPS_ARM64_ATOMICS) != 0);
-         if (have_rdm != have_atomics)
-            invalid_hwcaps(arch, hwcaps,
-                    "Mismatch detected between RDMA and atomics features.\n");
          return;
       }
 
