--- a/gdbserver_tests/filter_stderr       (revision 15586)
+++ b/gdbserver_tests/filter_stderr       (working copy)
@@ -10,4 +10,5 @@
     -e '/\/path\/to\/gdb/d' \
     -e '/and then give GDB the following command/d' \
     -e '/target remote |/d' \
-    -e '/pid is optional if only one valgrind process is running/d'
+    -e '/pid is optional if only one valgrind process is running/d' \
+    -e '/warning: remote target does not support file transfer, attempting to access files from local filesystem./d'


