commit a7216a4c4cc676e8f4750e71e7de654c0ce78ca3
Author: Will Schmidt <will_schmidt@vnet.ibm.com>
Date:   Tue Aug 24 13:18:11 2021 -0500

    Remove a unneeded / unnecessary prefix check.

diff --git a/VEX/priv/guest_ppc_toIR.c b/VEX/priv/guest_ppc_toIR.c
index f223fe9456..57ac7bcf48 100644
--- a/VEX/priv/guest_ppc_toIR.c
+++ b/VEX/priv/guest_ppc_toIR.c
@@ -14561,16 +14561,10 @@ static Bool dis_fp_pair_prefix ( UInt prefix, UInt theInstr )
 
       /* Endian aware prefixed load */
       pDIP( is_prefix, "stxvp %u,%llu(%u)\n", XTp, immediate_val, rA_addr );
       DIPp( is_prefix, ",%u", R );
 
-      if ( R == 1 ) {
-         vex_printf("Illegal instruction R = 1; pstxvp %u,%llu(%u)\n",
-                    XTp, immediate_val, rA_addr );
-         return False;
-      }
-
       assign( EA_8, binop( Iop_Add64, mkU64( 8 ), mkexpr( EA ) ) );
       assign( EA_16, binop( Iop_Add64, mkU64( 16 ), mkexpr( EA ) ) );
       assign( EA_24, binop( Iop_Add64, mkU64( 24 ), mkexpr( EA ) ) );
 
       if (host_endness == VexEndnessBE) {
