commit e5ff7e633a768a421dafbc115c3a3c2f713338db
Author: Mark Wielaard <mark@klomp.org>
Date:   Wed Apr 10 00:33:26 2019 +0200

    add missing a.c

diff --git a/cachegrind/tests/a.c b/cachegrind/tests/a.c
new file mode 100644
index 0000000..cd1c66e
--- /dev/null
+++ b/cachegrind/tests/a.c
@@ -0,0 +1,7 @@
+int main(void) {
+   int z = 0;
+   for (int i = 0; i < 1000000; i++) {
+      z += i;
+   }
+   return z % 256;
+}
