commit 983cb6ab1d60c79a8a9c6723815468fc59c4f248
Author: Mark Wielaard <mark@klomp.org>
Date:   Wed Apr 10 01:28:24 2019 +0200

    Bug 406357

diff --git a/gdbserver_tests/filter_gdb b/gdbserver_tests/filter_gdb
index 05bbe4c..d0ab11d 100755
--- a/gdbserver_tests/filter_gdb
+++ b/gdbserver_tests/filter_gdb
@@ -92,7 +92,7 @@ sed -e '/Remote debugging using/,/vgdb launched process attached/d'
     -e 's/^Thread .... received /Program received /'                                                  \
     -e 's/#[0-9]\(  0x........ in sleeper_or_burner\)/#.\1/'                                          \
     -e 's/\(#0  0x........ in do_burn ()\) at sleepers.c:41/\1/'                                      \
-    -e '/^Reading symbols from .*\.\.\.done\./d'                                                      \
+    -e '/^Reading symbols from .*\.\.\.\(done\.\)\?/d'                                                \
     -e '/^Loaded symbols for .*$/d'                                                                   \
     -e '/^Current language.*/d'                                                                       \
     -e '/^The current source language is.*/d'                                                         \
