diff --git a/exp-bbv/tests/amd64-linux/Makefile.am b/exp-bbv/tests/amd64-linux/Makefile.am
index d3eddb3..7ff8966 100644
--- a/exp-bbv/tests/amd64-linux/Makefile.am
+++ b/exp-bbv/tests/amd64-linux/Makefile.am
@@ -25,7 +25,7 @@ EXTRA_DIST = \
 
 AM_CCASFLAGS += -ffreestanding
 
-LDFLAGS += -nostartfiles -nodefaultlibs
+AM_LDFLAGS = -nostartfiles -nodefaultlibs
 
 clone_test_SOURCES = clone_test.S
 complex_rep_SOURCES = complex_rep.S
diff --git a/exp-bbv/tests/arm-linux/Makefile.am b/exp-bbv/tests/arm-linux/Makefile.am
index 9b91b28..d06b2d2 100644
--- a/exp-bbv/tests/arm-linux/Makefile.am
+++ b/exp-bbv/tests/arm-linux/Makefile.am
@@ -16,7 +16,7 @@ EXTRA_DIST = \
 
 AM_CCASFLAGS += -ffreestanding -Xassembler -I$(top_srcdir)/exp-bbv/tests
 
-LDFLAGS += -nostartfiles -nodefaultlibs
+AM_LDFLAGS = -nostartfiles -nodefaultlibs
 
 ll_SOURCES = ll.S
 million_SOURCES = million.S
diff --git a/exp-bbv/tests/ppc32-linux/Makefile.am b/exp-bbv/tests/ppc32-linux/Makefile.am
index 56646f4..691d803 100644
--- a/exp-bbv/tests/ppc32-linux/Makefile.am
+++ b/exp-bbv/tests/ppc32-linux/Makefile.am
@@ -16,7 +16,7 @@ EXTRA_DIST = \
 
 AM_CCASFLAGS += -ffreestanding -m32 -Xassembler -I$(top_srcdir)/exp-bbv/tests
 
-LDFLAGS += -nostartfiles -nodefaultlibs -m32
+AM_LDFLAGS = -nostartfiles -nodefaultlibs -m32
 
 ll_SOURCES = ll.S
 million_SOURCES = million.S
diff --git a/exp-bbv/tests/x86-linux/Makefile.am b/exp-bbv/tests/x86-linux/Makefile.am
index a16061f..96bb562 100644
--- a/exp-bbv/tests/x86-linux/Makefile.am
+++ b/exp-bbv/tests/x86-linux/Makefile.am
@@ -16,7 +16,7 @@ EXTRA_DIST = \
 
 AM_CCASFLAGS += -ffreestanding -Xassembler -I$(top_srcdir)/exp-bbv/tests
 
-LDFLAGS += @FLAG_M32@ -static -nostartfiles -nodefaultlibs
+AM_LDFLAGS = @FLAG_M32@ -static -nostartfiles -nodefaultlibs
 
 clone_test_SOURCES = clone_test.S
 ll_SOURCES = ll.S
diff --git a/exp-bbv/tests/x86/Makefile.am b/exp-bbv/tests/x86/Makefile.am
index 3857aae..d8be880 100644
--- a/exp-bbv/tests/x86/Makefile.am
+++ b/exp-bbv/tests/x86/Makefile.am
@@ -18,7 +18,7 @@ EXTRA_DIST = \
 
 AM_CCASFLAGS += -ffreestanding
 
-LDFLAGS += @FLAG_M32@ -static -nostartfiles -nodefaultlibs
+AM_LDFLAGS = @FLAG_M32@ -static -nostartfiles -nodefaultlibs
 
 complex_rep_SOURCES = complex_rep.S
 fldcw_check_SOURCES = fldcw_check.S
Only in valgrind-3.12.0.BETA1: autom4te.cache
diff -ru valgrind-3.12.0.BETA1.orig/exp-bbv/tests/amd64-linux/Makefile.in valgrind-3.12.0.BETA1/exp-bbv/tests/amd64-linux/Makefile.in
--- valgrind-3.12.0.BETA1.orig/exp-bbv/tests/amd64-linux/Makefile.in	2016-10-01 11:56:11.345194588 +0200
+++ valgrind-3.12.0.BETA1/exp-bbv/tests/amd64-linux/Makefile.in	2016-10-01 11:56:35.620748043 +0200
@@ -310,7 +310,7 @@
 INSTALL_PROGRAM = @INSTALL_PROGRAM@
 INSTALL_SCRIPT = @INSTALL_SCRIPT@
 INSTALL_STRIP_PROGRAM = @INSTALL_STRIP_PROGRAM@
-LDFLAGS = @LDFLAGS@ -nostartfiles -nodefaultlibs
+LDFLAGS = @LDFLAGS@
 LDFLAGS_MPI = @LDFLAGS_MPI@
 LIBOBJS = @LIBOBJS@
 LIBS = @LIBS@
@@ -610,6 +610,7 @@
 	   rep_prefix.stderr.exp \
 	   rep_prefix.vgtest 
 
+AM_LDFLAGS = -nostartfiles -nodefaultlibs
 clone_test_SOURCES = clone_test.S
 complex_rep_SOURCES = complex_rep.S
 fldcw_check_SOURCES = fldcw_check.S
diff -ru valgrind-3.12.0.BETA1.orig/exp-bbv/tests/arm-linux/Makefile.in valgrind-3.12.0.BETA1/exp-bbv/tests/arm-linux/Makefile.in
--- valgrind-3.12.0.BETA1.orig/exp-bbv/tests/arm-linux/Makefile.in	2016-10-01 11:56:11.344194606 +0200
+++ valgrind-3.12.0.BETA1/exp-bbv/tests/arm-linux/Makefile.in	2016-10-01 11:56:35.659747325 +0200
@@ -285,7 +285,7 @@
 INSTALL_PROGRAM = @INSTALL_PROGRAM@
 INSTALL_SCRIPT = @INSTALL_SCRIPT@
 INSTALL_STRIP_PROGRAM = @INSTALL_STRIP_PROGRAM@
-LDFLAGS = @LDFLAGS@ -nostartfiles -nodefaultlibs
+LDFLAGS = @LDFLAGS@
 LDFLAGS_MPI = @LDFLAGS_MPI@
 LIBOBJS = @LIBOBJS@
 LIBS = @LIBS@
@@ -577,6 +577,7 @@
 	   million.post.exp \
 	   million.vgtest
 
+AM_LDFLAGS = -nostartfiles -nodefaultlibs
 ll_SOURCES = ll.S
 million_SOURCES = million.S
 all: all-am
diff -ru valgrind-3.12.0.BETA1.orig/exp-bbv/tests/ppc32-linux/Makefile.in valgrind-3.12.0.BETA1/exp-bbv/tests/ppc32-linux/Makefile.in
--- valgrind-3.12.0.BETA1.orig/exp-bbv/tests/ppc32-linux/Makefile.in	2016-10-01 11:56:11.344194606 +0200
+++ valgrind-3.12.0.BETA1/exp-bbv/tests/ppc32-linux/Makefile.in	2016-10-01 11:56:35.698746608 +0200
@@ -285,7 +285,7 @@
 INSTALL_PROGRAM = @INSTALL_PROGRAM@
 INSTALL_SCRIPT = @INSTALL_SCRIPT@
 INSTALL_STRIP_PROGRAM = @INSTALL_STRIP_PROGRAM@
-LDFLAGS = @LDFLAGS@ -nostartfiles -nodefaultlibs -m32
+LDFLAGS = @LDFLAGS@
 LDFLAGS_MPI = @LDFLAGS_MPI@
 LIBOBJS = @LIBOBJS@
 LIBS = @LIBS@
@@ -577,6 +577,7 @@
 	   million.post.exp \
 	   million.vgtest
 
+AM_LDFLAGS = -nostartfiles -nodefaultlibs -m32
 ll_SOURCES = ll.S
 million_SOURCES = million.S
 all: all-am
diff -ru valgrind-3.12.0.BETA1.orig/exp-bbv/tests/x86/Makefile.in valgrind-3.12.0.BETA1/exp-bbv/tests/x86/Makefile.in
--- valgrind-3.12.0.BETA1.orig/exp-bbv/tests/x86/Makefile.in	2016-10-01 11:56:11.345194588 +0200
+++ valgrind-3.12.0.BETA1/exp-bbv/tests/x86/Makefile.in	2016-10-01 11:56:35.777745155 +0200
@@ -294,7 +294,7 @@
 INSTALL_PROGRAM = @INSTALL_PROGRAM@
 INSTALL_SCRIPT = @INSTALL_SCRIPT@
 INSTALL_STRIP_PROGRAM = @INSTALL_STRIP_PROGRAM@
-LDFLAGS = @LDFLAGS@ @FLAG_M32@ -static -nostartfiles -nodefaultlibs
+LDFLAGS = @LDFLAGS@
 LDFLAGS_MPI = @LDFLAGS_MPI@
 LIBOBJS = @LIBOBJS@
 LIBS = @LIBS@
@@ -587,6 +587,7 @@
 	   rep_prefix.stderr.exp \
 	   rep_prefix.vgtest 
 
+AM_LDFLAGS = @FLAG_M32@ -static -nostartfiles -nodefaultlibs
 complex_rep_SOURCES = complex_rep.S
 fldcw_check_SOURCES = fldcw_check.S
 million_SOURCES = million.S
diff -ru valgrind-3.12.0.BETA1.orig/exp-bbv/tests/x86-linux/Makefile.in valgrind-3.12.0.BETA1/exp-bbv/tests/x86-linux/Makefile.in
--- valgrind-3.12.0.BETA1.orig/exp-bbv/tests/x86-linux/Makefile.in	2016-10-01 11:56:11.344194606 +0200
+++ valgrind-3.12.0.BETA1/exp-bbv/tests/x86-linux/Makefile.in	2016-10-01 11:56:35.737745891 +0200
@@ -285,7 +285,7 @@
 INSTALL_PROGRAM = @INSTALL_PROGRAM@
 INSTALL_SCRIPT = @INSTALL_SCRIPT@
 INSTALL_STRIP_PROGRAM = @INSTALL_STRIP_PROGRAM@
-LDFLAGS = @LDFLAGS@ @FLAG_M32@ -static -nostartfiles -nodefaultlibs
+LDFLAGS = @LDFLAGS@
 LDFLAGS_MPI = @LDFLAGS_MPI@
 LIBOBJS = @LIBOBJS@
 LIBS = @LIBS@
@@ -577,6 +577,7 @@
 	   ll.post.exp \
 	   ll.vgtest
 
+AM_LDFLAGS = @FLAG_M32@ -static -nostartfiles -nodefaultlibs
 clone_test_SOURCES = clone_test.S
 ll_SOURCES = ll.S
 all: all-am
