Disable some hanging tests.

--- valgrind-3.8.0/helgrind/tests/Makefile.am.jj	2011-10-26 23:24:25.000000000 +0200
+++ valgrind-3.8.0/helgrind/tests/Makefile.am	2012-01-27 15:01:58.955624195 +0100
@@ -96,7 +96,6 @@ EXTRA_DIST = \
 # XXX: tc18_semabuse uses operations that are unsupported on Darwin.  It
 # should be conditionally compiled like tc20_verifywrap is.
 check_PROGRAMS = \
-	annotate_hbefore \
 	cond_timedwait_invalid \
 	free_is_write \
 	hg01_all_ok \
--- valgrind-3.8.0/helgrind/tests/Makefile.in.jj	2011-11-05 12:26:51.000000000 +0100
+++ valgrind-3.8.0/helgrind/tests/Makefile.in	2012-01-27 15:02:40.285383038 +0100
@@ -54,8 +54,7 @@
 @VGCONF_HAVE_PLATFORM_SEC_TRUE@am__append_2 = -DVGA_SEC_@VGCONF_ARCH_SEC@=1  \
 @VGCONF_HAVE_PLATFORM_SEC_TRUE@		-DVGP_SEC_@VGCONF_ARCH_PRI@_@VGCONF_OS@=1
 
-check_PROGRAMS = annotate_hbefore$(EXEEXT) \
-	cond_timedwait_invalid$(EXEEXT) free_is_write$(EXEEXT) \
+check_PROGRAMS = cond_timedwait_invalid$(EXEEXT) free_is_write$(EXEEXT) \
 	hg01_all_ok$(EXEEXT) hg02_deadlock$(EXEEXT) \
 	hg03_inherit$(EXEEXT) hg04_race$(EXEEXT) hg05_race2$(EXEEXT) \
 	hg06_readshared$(EXEEXT) locked_vs_unlocked1$(EXEEXT) \
